pkg/sdkserver/sdkserver.go

```diff
@@ -762,20 +784,196 @@ func (s *SDKServer) GetCounter(ctx context.Context, in *alpha.GetCounterRequest)
 	if !runtime.FeatureEnabled(runtime.FeatureCountsAndLists) {
 		return nil, errors.Errorf("%s not enabled", runtime.FeatureCountsAndLists)
 	}
-	// TODO(#2716): Implement me
-	return nil, errors.Errorf("Unimplemented -- GetCounter coming soon")
+
+	s.logger.WithField("name", in.Name).Info("Getting Counter")
+	s.gsUpdateMutex.RLock()
+	defer s.gsUpdateMutex.RUnlock()
+
+	// If we have made updates then return the SDK view of the Counter, otherwise get the counter from the gameserver
+	if counterUpdated, ok := s.gsCounterUpdates[in.Name]; ok {
+		s.logger.WithField("Get Counter from SDK Server", counterUpdated).Debugf("Got Counter %s", in.Name)
+		return &alpha.Counter{Name: in.Name, Count: counterUpdated.count, Capacity: counterUpdated.capacity}, nil
+	}
+
+	gs, err := s.gameServer()
+	if err != nil {
+		return nil, err
+	}
+
+	if counter, ok := gs.Status.Counters[in.Name]; ok {
+		s.logger.WithField("Get Counter", counter).Debugf("Got Counter %s", in.Name)
+		return &alpha.Counter{Name: in.Name, Count: counter.Count, Capacity: counter.Capacity}, nil
+	}
+
+	return nil, errors.Errorf("NOT_FOUND. %s Counter not found", in.Name)
 }
 
-// UpdateCounter returns the updated Counter. Returns NOT_FOUND if the Counter does not exist.
-// Returns INVALID_ARGUMENT if the FieldMask paths are invalid.
+// UpdateCounter collapses all UpdateCounterRequests for a given Counter into a single request.
+// Returns NOT_FOUND if the Counter does not exist (name cannot be updated).
+// Returns OUT_OF_RANGE if the Count is out of range [0,Capacity].
 // [Stage:Alpha]
 // [FeatureFlag:CountsAndLists]
 func (s *SDKServer) UpdateCounter(ctx context.Context, in *alpha.UpdateCounterRequest) (*alpha.Counter, error) {
 	if !runtime.FeatureEnabled(runtime.FeatureCountsAndLists) {
 		return nil, errors.Errorf("%s not enabled", runtime.FeatureCountsAndLists)
 	}
-	// TODO(#2716): Implement Me
-	return nil, errors.Errorf("Unimplemented -- UpdateCounter coming soon")
+
+	if in.CounterUpdateRequest == nil {
+		return nil, errors.Errorf("INVALID_ARGUMENT. CounterUpdateRequest: %v cannot be nil", in.CounterUpdateRequest)
+	}
+
+	s.logger.WithField("name", in.CounterUpdateRequest.Name).Info("Update Counter Request")
+	s.gsUpdateMutex.Lock()
+	defer s.gsUpdateMutex.Unlock()
+
+	// Check if we already have a batch request started for this Counter. If not, add new request to
+	// the gsCounterUpdates map.
+	name := in.CounterUpdateRequest.Name
+	batchCounter, batchOk := s.gsCounterUpdates[name]
+
+	if !batchOk {
+		gs, err := s.gameServer()
```

Comment:
I think to simplify things, we should be able to use `GetCounter()` to get what should be the latest information rather than rebuilding it here (although there is some trickiness with the locking - we may need a private `getCount()` function that doesn't have the lock, and make the calling function invoke it locked.)

Then you know that from `GetCounter()` you will get whatever the latest view is of the counts and lists, it should hopefully be easier to do validation without having to go check as much of the batching state.